File: android/app/src/main/kotlin/com/hadithly/app/core/data/ReadingPosition.kt
package com.hadithly.app.core.data

import kotlinx.serialization.Serializable
import java.security.MessageDigest
import kotlin.math.abs
import kotlin.math.roundToInt

@Serializable
data class ReadingPositionAnchor(
    val provider: String,
    val collectionSlug: String,
    val providerHadithId: String,
)

@Serializable
data class ReadingPosition(
    val schemaVersion: Double = SCHEMA_VERSION,
    val anchor: ReadingPositionAnchor,
    val contentVersion: String,
    val volumeId: String,
    val chapterId: String? = null,
    val pageKey: String,
    val displayPageIndex: Double,
    val rawPageOffset: Double,
    val normalizedOffset: Double,
    val layoutSignature: String,
    val updatedAt: Double,
) {
    init {
        require(schemaVersion == SCHEMA_VERSION)
        require(anchor.provider in setOf("sunnah_now", "sunnah_com", "local_dump"))
        require(anchor.collectionSlug.matches(Regex("^[a-z][a-z0-9-]{1,31}$")))
        require(anchor.providerHadithId.matches(Regex("^[0-9]{1,12}(\\.[0-9]{1,12})*$")))
        require(contentVersion.matches(Regex("^(cv1:[A-Za-z0-9._:-]{1,128}|legacy)$")))
        require(volumeId.matches(Regex("^[A-Za-z0-9._:-]{1,128}$")))
        require(chapterId == null || chapterId.matches(Regex("^[A-Za-z0-9._:-]{1,128}$")))
        require(pageKey.matches(Regex("^(pg1:[A-Za-z0-9._:-]{1,160}|legacy)$")))
        require(displayPageIndex.isFinite() && displayPageIndex >= 1.0 && displayPageIndex % 1.0 == 0.0)
        require(rawPageOffset.isFinite() && rawPageOffset >= 0.0)
        require(normalizedOffset.isFinite() && normalizedOffset in 0.0..1.0)
        require(layoutSignature.matches(Regex("^(ls1:[A-Za-z0-9._:-]{1,160}|legacy)$")))
        require(updatedAt.isFinite() && updatedAt >= 0.0)
    }

    companion object {
        const val SCHEMA_VERSION = 1.0

        fun legacy(
            collectionSlug: String,
            providerHadithId: String,
            volumeId: String,
            updatedAt: Double,
            provider: String = "sunnah_now",
            chapterId: String? = null,
        ): ReadingPosition? {
            if (!collectionSlug.matches(Regex("^[a-z][a-z0-9-]{1,31}$"))) return null
            if (!providerHadithId.matches(Regex("^[0-9]{1,12}(\\.[0-9]{1,12})*$"))) return null
            if (volumeId.isBlank()) return null
            return ReadingPosition(
                anchor = ReadingPositionAnchor(provider, collectionSlug, providerHadithId),
                contentVersion = "legacy",
                volumeId = volumeId,
                chapterId = chapterId,
                pageKey = "legacy",
                displayPageIndex = 1.0,
                rawPageOffset = 0.0,
                normalizedOffset = 0.0,
                layoutSignature = "legacy",
                updatedAt = updatedAt,
            )
        }
    }
}

enum class ReadingWidthClass(val wire: String) {
    COMPACT("compact"),
    REGULAR("regular");

    companion object {
        fun fromWidthDp(widthDp: Double): ReadingWidthClass =
            if (widthDp < 600.0) COMPACT else REGULAR
    }
}

data class ReadingLayout(
    val locale: String,
    val arabicVisible: Boolean,
    val translationVisible: Boolean,
    val arabicFontId: String,
    val arabicFontSize: Double,
    val widthClass: ReadingWidthClass,
    val paginationVersion: Double,
) {
    val signature: String
        get() {
            val canonical = listOf(
                "v1",
                "locale=$locale",
                "arabic=${if (arabicVisible) 1 else 0}",
                "translation=${if (translationVisible) 1 else 0}",
                "font=$arabicFontId",
                "size=${(arabicFontSize * 1_000).roundToInt()}",
                "width=${widthClass.wire}",
                "pagination=${paginationVersion.toInt()}",
            ).joinToString("|")
            val digest = MessageDigest.getInstance("SHA-256").digest(canonical.toByteArray(Charsets.UTF_8))
            return "ls1:" + digest.joinToString("") { "%02x".format(it.toInt() and 0xff) }
        }
}

data class ReadingPositionAnchorLocation(
    val anchor: ReadingPositionAnchor,
    val volumeId: String,
    val chapterId: String?,
)

data class ReadingPositionPage(
    val pageKey: String,
    val displayPageIndex: Double,
    val anchors: List<ReadingPositionAnchorLocation>,
)

data class ReadingPositionTopology(
    val contentVersion: String,
    val layoutSignature: String,
    val pages: List<ReadingPositionPage>,
)

object ReadingPositionResolver {
    enum class Kind { RAW, SEMANTIC, NEAREST, UNAVAILABLE }

    data class Result(
        val kind: Kind,
        val anchor: ReadingPositionAnchor?,
        val volumeId: String?,
        val chapterId: String?,
        val pageKey: String?,
        val displayPageIndex: Double,
        val rawPageOffset: Double,
        val normalizedOffset: Double,
    )

    private data class Located(
        val location: ReadingPositionAnchorLocation,
        val pageKey: String,
        val displayPageIndex: Double,
        val itemIndex: Int,
    )

    fun resolve(position: ReadingPosition, topology: ReadingPositionTopology): Result {
        val candidates = topology.pages.flatMap { page ->
            page.anchors.mapIndexed { itemIndex, anchor ->
                Located(anchor, page.pageKey, page.displayPageIndex, itemIndex)
            }
        }.filter {
            it.location.anchor.provider == position.anchor.provider &&
                it.location.anchor.collectionSlug == position.anchor.collectionSlug
        }
        val exact = candidates.firstOrNull { it.location.anchor == position.anchor }
        if (exact != null) {
            val rawMatches = topology.contentVersion == position.contentVersion &&
                topology.layoutSignature == position.layoutSignature &&
                exact.pageKey == position.pageKey &&
                exact.displayPageIndex == position.displayPageIndex
            return result(
                if (rawMatches) Kind.RAW else Kind.SEMANTIC,
                exact,
                if (rawMatches) position.rawPageOffset else 0.0,
                position.normalizedOffset,
            )
        }
        if (candidates.isEmpty()) {
            return Result(Kind.UNAVAILABLE, null, null, null, null, 1.0, 0.0, 0.0)
        }
        var preferred = candidates
        val sameVolume = preferred.filter { it.location.volumeId == position.volumeId }
        if (sameVolume.isNotEmpty()) preferred = sameVolume
        position.chapterId?.let { chapterId ->
            val sameChapter = preferred.filter { it.location.chapterId == chapterId }
            if (sameChapter.isNotEmpty()) preferred = sameChapter
        }
        val nearest = preferred.minWithOrNull { left, right ->
            compareDistance(position.anchor.providerHadithId, left, right)
        } ?: return Result(Kind.UNAVAILABLE, null, null, null, null, 1.0, 0.0, 0.0)
        return result(Kind.NEAREST, nearest, 0.0, 0.0)
    }

    private fun result(kind: Kind, located: Located, raw: Double, normalized: Double) = Result(
        kind = kind,
        anchor = located.location.anchor,
        volumeId = located.location.volumeId,
        chapterId = located.location.chapterId,
        pageKey = located.pageKey,
        displayPageIndex = located.displayPageIndex,
        rawPageOffset = raw,
        normalizedOffset = normalized,
    )

    private fun compareDistance(targetId: String, left: Located, right: Located): Int {
        val target = segments(targetId) ?: return compareLocated(left, right)
        val leftParts = segments(left.location.anchor.providerHadithId)
        val rightParts = segments(right.location.anchor.providerHadithId)
        if (leftParts == null || rightParts == null) {
            if (leftParts != null) return -1
            if (rightParts != null) return 1
            return compareLocated(left, right)
        }
        val width = maxOf(target.size, leftParts.size, rightParts.size)
        for (index in 0 until width) {
            val targetPart = target.getOrElse(index) { 0L }
            val leftDistance = abs(leftParts.getOrElse(index) { 0L } - targetPart)
            val rightDistance = abs(rightParts.getOrElse(index) { 0L } - targetPart)
            if (leftDistance != rightDistance) return leftDistance.compareTo(rightDistance)
        }
        val providerOrder = compareSegments(leftParts, rightParts)
        if (providerOrder != 0) return providerOrder
        return compareLocated(left, right)
    }

    private fun compareLocated(left: Located, right: Located): Int {
        val idOrder = left.location.anchor.providerHadithId.compareTo(right.location.anchor.providerHadithId)
        if (idOrder != 0) return idOrder
        val pageOrder = left.displayPageIndex.compareTo(right.displayPageIndex)
        return if (pageOrder != 0) pageOrder else left.itemIndex.compareTo(right.itemIndex)
    }

    private fun segments(value: String): List<Long>? {
        val result = value.split('.').map(String::toLongOrNull)
        if (result.any { it == null }) return null
        return result.filterNotNull()
    }

    private fun compareSegments(left: List<Long>, right: List<Long>): Int {
        for (index in 0 until maxOf(left.size, right.size)) {
            val result = left.getOrElse(index) { 0L }.compareTo(right.getOrElse(index) { 0L })
            if (result != 0) return result
        }
        return left.size.compareTo(right.size)
    }
}

fun ReadingPosition.toWireMap(): Map<String, Any?> = buildMap {
    put("schemaVersion", schemaVersion)
    put("anchor", mapOf(
        "provider" to anchor.provider,
        "collectionSlug" to anchor.collectionSlug,
        "providerHadithId" to anchor.providerHadithId,
    ))
    put("contentVersion", contentVersion)
    put("volumeId", volumeId)
    chapterId?.let { put("chapterId", it) }
    put("pageKey", pageKey)
    put("displayPageIndex", displayPageIndex)
    put("rawPageOffset", rawPageOffset)
    put("normalizedOffset", normalizedOffset)
    put("layoutSignature", layoutSignature)
    put("updatedAt", updatedAt)
}

File: ios/Hadithly/Core/ReadingPosition/ReadingPosition.swift
import CryptoKit
import Foundation

struct ReadingPositionAnchor: Codable, Equatable, Sendable {
    let provider: String
    let collectionSlug: String
    let providerHadithId: String
}

struct ReadingPosition: Codable, Equatable, Sendable {
    static let currentSchemaVersion = 1.0

    let schemaVersion: Double
    let anchor: ReadingPositionAnchor
    let contentVersion: String
    let volumeId: String
    let chapterId: String?
    let pageKey: String
    let displayPageIndex: Double
    let rawPageOffset: Double
    let normalizedOffset: Double
    let layoutSignature: String
    let updatedAt: Double

    init(
        anchor: ReadingPositionAnchor,
        contentVersion: String,
        volumeId: String,
        chapterId: String? = nil,
        pageKey: String,
        displayPageIndex: Double,
        rawPageOffset: Double,
        normalizedOffset: Double,
        layoutSignature: String,
        updatedAt: Double
    ) {
        self.schemaVersion = Self.currentSchemaVersion
        self.anchor = anchor
        self.contentVersion = contentVersion
        self.volumeId = volumeId
        self.chapterId = chapterId
        self.pageKey = pageKey
        self.displayPageIndex = displayPageIndex
        self.rawPageOffset = rawPageOffset
        self.normalizedOffset = normalizedOffset
        self.layoutSignature = layoutSignature
        self.updatedAt = updatedAt
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        schemaVersion = try container.decode(Double.self, forKey: .schemaVersion)
        anchor = try container.decode(ReadingPositionAnchor.self, forKey: .anchor)
        contentVersion = try container.decode(String.self, forKey: .contentVersion)
        volumeId = try container.decode(String.self, forKey: .volumeId)
        chapterId = try container.decodeIfPresent(String.self, forKey: .chapterId)
        pageKey = try container.decode(String.self, forKey: .pageKey)
        displayPageIndex = try container.decode(Double.self, forKey: .displayPageIndex)
        rawPageOffset = try container.decode(Double.self, forKey: .rawPageOffset)
        normalizedOffset = try container.decode(Double.self, forKey: .normalizedOffset)
        layoutSignature = try container.decode(String.self, forKey: .layoutSignature)
        updatedAt = try container.decode(Double.self, forKey: .updatedAt)
        guard schemaVersion == Self.currentSchemaVersion,
              ["sunnah_now", "sunnah_com", "local_dump"].contains(anchor.provider),
              ReadingPositionValidation.validCollectionSlug(anchor.collectionSlug),
              ReadingPositionValidation.validProviderHadithId(anchor.providerHadithId),
              ReadingPositionValidation.validContentVersion(contentVersion),
              ReadingPositionValidation.validContextId(volumeId),
              chapterId.map(ReadingPositionValidation.validContextId) ?? true,
              ReadingPositionValidation.validPageKey(pageKey),
              displayPageIndex.isFinite,
              displayPageIndex >= 1,
              displayPageIndex.rounded() == displayPageIndex,
              rawPageOffset.isFinite,
              rawPageOffset >= 0,
              normalizedOffset.isFinite,
              (0...1).contains(normalizedOffset),
              ReadingPositionValidation.validLayoutSignature(layoutSignature),
              updatedAt.isFinite,
              updatedAt >= 0 else {
            throw DecodingError.dataCorrupted(
                .init(codingPath: decoder.codingPath, debugDescription: "Invalid ReadingPosition V1")
            )
        }
    }

    static func legacy(
        provider: String = "sunnah_now",
        collectionSlug: String,
        providerHadithId: String,
        volumeId: String,
        chapterId: String? = nil,
        updatedAt: Double
    ) -> ReadingPosition? {
        guard ReadingPositionValidation.validCollectionSlug(collectionSlug),
              ReadingPositionValidation.validProviderHadithId(providerHadithId),
              !volumeId.isEmpty else { return nil }
        return ReadingPosition(
            anchor: .init(
                provider: provider,
                collectionSlug: collectionSlug,
                providerHadithId: providerHadithId
            ),
            contentVersion: "legacy",
            volumeId: volumeId,
            chapterId: chapterId,
            pageKey: "legacy",
            displayPageIndex: 1,
            rawPageOffset: 0,
            normalizedOffset: 0,
            layoutSignature: "legacy",
            updatedAt: updatedAt
        )
    }
}

private enum ReadingPositionValidation {
    static func validCollectionSlug(_ value: String) -> Bool {
        value.range(of: "^[a-z][a-z0-9-]{1,31}$", options: .regularExpression) != nil
    }

    static func validProviderHadithId(_ value: String) -> Bool {
        value.range(of: "^[0-9]{1,12}(\\.[0-9]{1,12})*$", options: .regularExpression) != nil
    }

    static func validContentVersion(_ value: String) -> Bool {
        value.range(of: "^(cv1:[A-Za-z0-9._:-]{1,128}|legacy)$", options: .regularExpression) != nil
    }

    static func validPageKey(_ value: String) -> Bool {
        value.range(of: "^(pg1:[A-Za-z0-9._:-]{1,160}|legacy)$", options: .regularExpression) != nil
    }

    static func validLayoutSignature(_ value: String) -> Bool {
        value.range(of: "^(ls1:[A-Za-z0-9._:-]{1,160}|legacy)$", options: .regularExpression) != nil
    }

    static func validContextId(_ value: String) -> Bool {
        value.range(of: "^[A-Za-z0-9._:-]{1,128}$", options: .regularExpression) != nil
    }
}

enum ReadingWidthClass: String, Codable, Sendable {
    case compact
    case regular
}

struct ReadingLayout: Equatable, Sendable {
    let locale: String
    let arabicVisible: Bool
    let translationVisible: Bool
    let arabicFontId: String
    let arabicFontSize: Double
    let widthClass: ReadingWidthClass
    let paginationVersion: Double

    var signature: String {
        let sizeMilliPoints = Int((arabicFontSize * 1_000).rounded())
        let pagination = Int(paginationVersion)
        let canonical = [
            "v1",
            "locale=\(locale)",
            "arabic=\(arabicVisible ? 1 : 0)",
            "translation=\(translationVisible ? 1 : 0)",
            "font=\(arabicFontId)",
            "size=\(sizeMilliPoints)",
            "width=\(widthClass.rawValue)",
            "pagination=\(pagination)",
        ].joined(separator: "|")
        let digest = SHA256.hash(data: Data(canonical.utf8))
        return "ls1:" + digest.map { String(format: "%02x", $0) }.joined()
    }

    func withPaginationVersion(_ version: Double) -> ReadingLayout {
        ReadingLayout(
            locale: locale,
            arabicVisible: arabicVisible,
            translationVisible: translationVisible,
            arabicFontId: arabicFontId,
            arabicFontSize: arabicFontSize,
            widthClass: widthClass,
            paginationVersion: version
        )
    }
}

struct ReadingPositionAnchorLocation: Equatable, Sendable {
    let anchor: ReadingPositionAnchor
    let volumeId: String
    let chapterId: String?
}

struct ReadingPositionPage: Equatable, Sendable {
    let pageKey: String
    let displayPageIndex: Double
    let anchors: [ReadingPositionAnchorLocation]
}

struct ReadingPositionTopology: Equatable, Sendable {
    let contentVersion: String
    let layoutSignature: String
    let pages: [ReadingPositionPage]
}

struct ResolvedReadingPosition: Equatable, Sendable {
    enum Kind: Equatable, Sendable {
        case raw
        case semantic
        case nearest
        case unavailable
    }

    let kind: Kind
    let anchor: ReadingPositionAnchor?
    let volumeId: String?
    let chapterId: String?
    let pageKey: String?
    let displayPageIndex: Double
    let rawPageOffset: Double
    let normalizedOffset: Double
}

enum ReadingPositionResolver {
    private struct LocatedAnchor {
        let location: ReadingPositionAnchorLocation
        let pageKey: String
        let displayPageIndex: Double
        let itemIndex: Int
    }

    static func resolve(
        _ position: ReadingPosition,
        topology: ReadingPositionTopology
    ) -> ResolvedReadingPosition {
        let all = topology.pages.flatMap { page in
            page.anchors.enumerated().map { itemIndex, anchor in
                LocatedAnchor(
                    location: anchor,
                    pageKey: page.pageKey,
                    displayPageIndex: page.displayPageIndex,
                    itemIndex: itemIndex
                )
            }
        }
        let candidates = all.filter {
            $0.location.anchor.provider == position.anchor.provider
                && $0.location.anchor.collectionSlug == position.anchor.collectionSlug
        }
        if let exact = candidates.first(where: { $0.location.anchor == position.anchor }) {
            let rawMatches = topology.contentVersion == position.contentVersion
                && topology.layoutSignature == position.layoutSignature
                && exact.pageKey == position.pageKey
                && exact.displayPageIndex == position.displayPageIndex
            return result(
                kind: rawMatches ? .raw : .semantic,
                location: exact,
                rawPageOffset: rawMatches ? position.rawPageOffset : 0,
                normalizedOffset: position.normalizedOffset
            )
        }
        guard !candidates.isEmpty else {
            return ResolvedReadingPosition(
                kind: .unavailable,
                anchor: nil,
                volumeId: nil,
                chapterId: nil,
                pageKey: nil,
                displayPageIndex: 1,
                rawPageOffset: 0,
                normalizedOffset: 0
            )
        }
        var preferred = candidates
        let sameVolume = preferred.filter { $0.location.volumeId == position.volumeId }
        if !sameVolume.isEmpty { preferred = sameVolume }
        if let chapterId = position.chapterId {
            let sameChapter = preferred.filter { $0.location.chapterId == chapterId }
            if !sameChapter.isEmpty { preferred = sameChapter }
        }
        guard let nearest = preferred.sorted(by: {
            compareDistance(target: position.anchor.providerHadithId, left: $0, right: $1) < 0
        }).first else {
            return ResolvedReadingPosition(
                kind: .unavailable,
                anchor: nil,
                volumeId: nil,
                chapterId: nil,
                pageKey: nil,
                displayPageIndex: 1,
                rawPageOffset: 0,
                normalizedOffset: 0
            )
        }
        return result(kind: .nearest, location: nearest, rawPageOffset: 0, normalizedOffset: 0)
    }

    private static func result(
        kind: ResolvedReadingPosition.Kind,
        location: LocatedAnchor,
        rawPageOffset: Double,
        normalizedOffset: Double
    ) -> ResolvedReadingPosition {
        ResolvedReadingPosition(
            kind: kind,
            anchor: location.location.anchor,
            volumeId: location.location.volumeId,
            chapterId: location.location.chapterId,
            pageKey: location.pageKey,
            displayPageIndex: location.displayPageIndex,
            rawPageOffset: rawPageOffset,
            normalizedOffset: normalizedOffset
        )
    }

    private static func compareDistance(
        target: String,
        left: LocatedAnchor,
        right: LocatedAnchor
    ) -> Int {
        guard let targetSegments = segments(target) else {
            return compareLocated(left, right)
        }
        let leftSegments = segments(left.location.anchor.providerHadithId)
        let rightSegments = segments(right.location.anchor.providerHadithId)
        if leftSegments == nil || rightSegments == nil {
            if leftSegments != nil { return -1 }
            if rightSegments != nil { return 1 }
            return compareLocated(left, right)
        }
        guard let leftSegments, let rightSegments else {
            return compareLocated(left, right)
        }
        let width = max(targetSegments.count, max(leftSegments.count, rightSegments.count))
        for index in 0..<width {
            let targetPart = index < targetSegments.count ? targetSegments[index] : 0
            let leftPart = index < leftSegments.count ? leftSegments[index] : 0
            let rightPart = index < rightSegments.count ? rightSegments[index] : 0
            let leftDistance = abs(leftPart - targetPart)
            let rightDistance = abs(rightPart - targetPart)
            if leftDistance != rightDistance { return leftDistance < rightDistance ? -1 : 1 }
        }
        let providerOrder = compareSegments(leftSegments, rightSegments)
        if providerOrder != 0 { return providerOrder }
        return compareLocated(left, right)
    }

    private static func compareLocated(_ left: LocatedAnchor, _ right: LocatedAnchor) -> Int {
        let leftId = left.location.anchor.providerHadithId
        let rightId = right.location.anchor.providerHadithId
        if leftId != rightId { return leftId < rightId ? -1 : 1 }
        if left.displayPageIndex != right.displayPageIndex {
            return left.displayPageIndex < right.displayPageIndex ? -1 : 1
        }
        return left.itemIndex - right.itemIndex
    }

    private static func segments(_ value: String) -> [Int64]? {
        let result = value.split(separator: ".").map { Int64($0) }
        guard result.allSatisfy({ $0 != nil }) else { return nil }
        return result.compactMap { $0 }
    }

    private static func compareSegments(_ left: [Int64], _ right: [Int64]) -> Int {
        let width = max(left.count, right.count)
        for index in 0..<width {
            let difference = (index < left.count ? left[index] : 0) - (index < right.count ? right[index] : 0)
            if difference != 0 { return difference < 0 ? -1 : 1 }
        }
        return left.count - right.count
    }
}
